### PR TITLE
WebView Fixes

### PR DESCRIPTION
**Pull request by dscrobonia**
## Summary

- Moves the injection script to a `WKUserScript` to be loaded via the `WKWebView` configuration with the `injectionTime: .atDocumentStart` property set to guarantee our script loads first. 
  ```
   let script = WKUserScript(source: scriptSource, injectionTime: .atDocumentStart, forMainFrameOnly: true)
  ```
- Refactors the log for initializing the webview into `InitWebView`.
- Removes the webview persistence by default so that each session is fresh, so that caching issues don't occur. (this does mean if a user manually connects it won't persist between sessions)
   ```
   configuration.websiteDataStore = WKWebsiteDataStore.default() // Allows for data persistence across sessions
   ```
- Adds a new constructor overload to manually set the `persistSessionData` property
- Adds a fix to open new tabs within the same WebView

### New Tab Before

https://github.com/portal-hq/PortalSwift/assets/5150944/f5572b1f-0a99-4728-8bf4-ffff4f421fab

### New Tab After

https://github.com/portal-hq/PortalSwift/assets/5150944/9eefd0b1-6957-474a-b9e5-33075fc5b4db

### Code
- Checked against coding style guide: No
  
  
- Documentation updated: No
  
  

### Impact
- Breaking Changes: No
  
  
- Performance impact: No
  

### Testing
- Unit tests added: No
  
- E2E tests added: No
  
- Manual tests in staging: Pending
  
- E2E tests in Datadog: N/A - but manually tested all 3 approved dApps for Swift.
  

### Misc.
- Metrics, logs, or traces added: No
- Changelog updated: No, I believe this happens on release?

### CODE REVIEW

File: Sources/PortalSwift/Components/PortalWebView.swift [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 import UIKit
 import WebKit
+//import SwiftUI
```
Comment (Rshahatit) [inline Sources/PortalSwift/Components/PortalWebView.swift:10]: do we need this?

Comment (dscrobonia) [reply Sources/PortalSwift/Components/PortalWebView.swift:10]: oh no good catch

Comment (dscrobonia) [reply Sources/PortalSwift/Components/PortalWebView.swift:10]: oh no good catch

File: Sources/PortalSwift/Components/PortalWebView.swift [hunk 2]
```diff
@@ -89,33 +80,35 @@ public class PortalWebView: UIViewController, WKNavigationDelegate, WKScriptMess
...
+      print("[PortalWebView] No address found for user. Cannot inject provider into web page.")
+      return
+    }
+    self.webView = initWebView(address: address)
+  }
+  
+  /// The constructor for Portal's WebViewController.
+  /// - Parameters:
+  ///   - portal: Your Portal instance.
+  ///   - url: The URL the web view should start at.
```
Comment (Rshahatit) [inline Sources/PortalSwift/Components/PortalWebView.swift:94]: add a description for the `persistSessionData` argument

File: Sources/PortalSwift/Components/PortalWebView.swift [hunk 3]
```diff
@@ -147,6 +141,54 @@ public class PortalWebView: UIViewController, WKNavigationDelegate, WKScriptMess
...
+       configuration.websiteDataStore = WKWebsiteDataStore.default()
+      }
+
+      let webView = WKWebView(frame: .zero, configuration: configuration)
+      webView.scrollView.bounces = false
+      webView.navigationDelegate = self
+      webView.uiDelegate = self
+
+
+      // Enable debugging the webview in Safari.
```
Comment (Rshahatit) [inline Sources/PortalSwift/Components/PortalWebView.swift:178]: maybe we can add a debug flag to the args for the init function. Im sure customers would want to use it as well

File: Sources/PortalSwift/Components/PortalWebView.swift [hunk 4]
```diff
@@ -177,29 +219,24 @@ public class PortalWebView: UIViewController, WKNavigationDelegate, WKScriptMess
...
-      apiKey: self.portal.apiKey,
-      chainId: String(self.portal.chainId),
-      gatewayConfig: self.portal.gatewayConfig[self.portal.chainId]!,
-      autoApprove: self.portal.autoApprove,
-      enableMpc: true
-    )
-
-    self.evaluateJavascript(javascript, sourceURL: "portal_sign")
-
     self.onPageComplete?()
```
Comment (Rshahatit) [inline Sources/PortalSwift/Components/PortalWebView.swift:226]: Did you move the code above to a function called onPageComplete? or how is the script getting injected now?

File: Sources/PortalSwift/Components/PortalWebView.swift [hunk 5]
```diff
@@ -177,29 +218,24 @@ public class PortalWebView: UIViewController, WKNavigationDelegate, WKScriptMess
...
-    let javascript = self.injectPortal(
-      address: address,
-      apiKey: self.portal.apiKey,
-      chainId: String(self.portal.chainId),
-      gatewayConfig: self.portal.gatewayConfig[self.portal.chainId]!,
-      autoApprove: self.portal.autoApprove,
-      enableMpc: true
-    )
-
-    self.evaluateJavascript(javascript, sourceURL: "portal_sign")
```
Comment (Rshahatit) [inline Sources/PortalSwift/Components/PortalWebView.swift:199]: where did this code go for evaluateJavascript?

File: Sources/PortalSwift/Components/PortalWebView.swift (hunk 4 above)
Comment (Rshahatit) [reply Sources/PortalSwift/Components/PortalWebView.swift:226]: i see you moved it to the init function for the injectPortal function call

Comment (dscrobonia) [reply Sources/PortalSwift/Components/PortalWebView.swift:226]: The injection script is moved to a WKUserScript to be loaded via the WKWebView configuration with the injectionTime: .atDocumentStart property set to guarantee our script loads first.

File: Sources/PortalSwift/Components/PortalWebView.swift (hunk 3 above)
Comment (dscrobonia) [reply Sources/PortalSwift/Components/PortalWebView.swift:177]: Its currently using compiler directives to determine whether to enable that flag. When we package the code, is it cool to leave compiler directives in?

File: Sources/PortalSwift/Components/PortalWebView.swift (hunk 5 above)
Comment (dscrobonia) [reply Sources/PortalSwift/Components/PortalWebView.swift:199]: ah this is also not necessary. It's handled by the comment below. The `portal_sign` is confusing. It's just the name of the script being injected

File: SPM Example/PortalSwift/Portal.swift [hunk 6]
```diff
@@ -112,7 +112,7 @@ class PortalWrapper {
         chainId: chainId,
         keychain: keychain,
         gatewayConfig: [
-          11_155_111: "https://eth-sepolia.g.alchemy.com/v2/\(ALCHEMY_API_KEY)", 1: "https://eth-mainnet.g.alchemy.com/v2/\(ALCHEMY_API_KEY)", 80001: "https://polygon-mumbai.g.alchemy.com/v2/\(ALCHEMY_API_KEY)", 137: "https://polygon-mainnet.g.alchemy.com/v2/\(ALCHEMY_API_KEY)",
+          11_155_111: "https://eth-sepolia.g.alchemy.com/v2/\(ALCHEMY_API_KEY)", 5: "https://goerli.g.alchemy.com/v2/\(ALCHEMY_API_KEY)",1: "https://eth-mainnet.g.alchemy.com/v2/\(ALCHEMY_API_KEY)", 80001: "https://polygon-mumbai.g.alchemy.com/v2/\(ALCHEMY_API_KEY)", 137: "https://polygon-mainnet.g.alchemy.com/v2/\(ALCHEMY_API_KEY)",
```
Comment (kelsonic) [inline SPM Example/PortalSwift/Portal.swift:115]: Nit: Missing space between `,` and `1:`

File: Sources/PortalSwift/Components/PortalWebView.swift [hunk 7]
```diff
@@ -44,7 +44,7 @@ public class PortalWebView: UIViewController, WKNavigationDelegate, WKScriptMess
   private var onError: (Result<Any>) -> Void
   private var onPageStart: (() -> Void)?
   private var onPageComplete: (() -> Void)?
-
+
```
Comment (kelsonic) [inline Sources/PortalSwift/Components/PortalWebView.swift:47]: Nit: Added whitespace here, is your linter working by chance? I'm happy to add a commit with the `swiftformat` changes if you want 🙂

File: Sources/PortalSwift/Components/PortalWebView.swift [hunk 8]
```diff
@@ -177,29 +219,24 @@ public class PortalWebView: UIViewController, WKNavigationDelegate, WKScriptMess
...
-
     self.onPageComplete?()
   }
+  
+  /// Called when a new tab is opened.
+  public func webView(_ webView: WKWebView, createWebViewWith configuration: WKWebViewConfiguration, for navigationAction: WKNavigationAction, windowFeatures: WKWindowFeatures) -> WKWebView? {
+          if navigationAction.targetFrame == nil {
+              webView.load(navigationAction.request)
+            
+              // open in Safari
```
Comment (kelsonic) [inline Sources/PortalSwift/Components/PortalWebView.swift:235]: Can you add a bit more to this comment when we'd want to open it in Safari?

**Review by kelsonic — APPROVED** [review]
LGTM! 🎉

File: Sources/PortalSwift/Components/PortalWebView.swift (hunk 7 above)
Comment (dscrobonia) [reply Sources/PortalSwift/Components/PortalWebView.swift:47]: Ah let me get that configured correctly and re-run it. Thanks!

Comment (dscrobonia) [reply Sources/PortalSwift/Components/PortalWebView.swift]: Done!

File: Sources/PortalSwift/Components/PortalWebView.swift (hunk 8 above)
Comment (dscrobonia) [reply Sources/PortalSwift/Components/PortalWebView.swift]: Updated comment!

**Pull request merged**